### PR TITLE
(feat) Migrating conditions form to use RHF and Zod

### PR DESCRIPTION
**Pull request by lumuchris256**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [x] My work includes tests or is validated by existing tests.

## Summary
 
   -migrated the conditions form to use RHF and Zod

## Screenshots

https://github.com/openmrs/openmrs-esm-patient-chart/assets/59338693/c6047af6-587c-4093-a8d6-2a8eb95d25f2

current update 

focus

## Related Issue

https://issues.openmrs.org/browse/O3-2139 current
needs merging first  https://github.com/openmrs/openmrs-esm-patient-chart/pull/1232

## Other
cc @hadijahkyampeire @denniskigen

### CODE REVIEW

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx [hunk 1]
```diff
@@ -1,59 +1,78 @@
...
+    resolver: zodResolver(conditionSchema),
+    defaultValues: {
+      onsetDateTime:
+        formContext == 'editing'
+          ? matchingCondition?.onsetDateTime
+            ? new Date(matchingCondition?.onsetDateTime)
+            : null
+          : null,
+      clinicalStatus: condition?.cells?.find((cell) => cell?.info?.header === 'clinicalStatus')?.value ?? 'Active',
+      search: condition?.cells?.find((cell) => cell?.info?.header === 'display')?.value,
```
Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx:48]: initilizing RHF  with default values

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx [hunk 2]
```diff
@@ -1,59 +1,78 @@
...
   patientUuid?: string;
 }
+const conditionSchema = z.object({
+  clinicalStatus: z.string(),
+  endDate: z.date().optional(),
+  onsetDateTime: z.date().nullable(),
+  search: z.string({ required_error: "A condition is required"}),
+});
+
+export type ConditionFormData = z.infer<typeof conditionSchema>;
```
Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx:25]: infering form data type from the zod schema

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 3]
```diff
@@ -176,34 +159,42 @@ const ConditionsWidget: React.FC<ConditionsWidgetProps> = ({
...
+  const searchInputFocus = () => {
+    searchInputRef.current.focus();
+  };
+
   useEffect(() => {
-    const subscription = submissionNotifier.subscribe(({ isSubmitting }) => {
-      if (isSubmitting) {
-        editing ? handleUpdate() : handleCreate();
+    if (formState?.errors?.search) {
+      searchInputFocus();
```
Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:187]: focuses on the conditon's search input if an invalid condition is entered

Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:187]: focuses on the conditon's search input if an invalid condition is entered

File: packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx [hunk 4]
```diff
@@ -1,104 +1,116 @@
...
-            {isSubmittingForm ? (
-              <InlineLoading description={t('saving', 'Saving') + '...'} />
-            ) : (
-              <span>{t('saveAndClose', 'Save & close')}</span>
-            )}
-          </Button>
-        </ButtonSet>
-      </div>
-    </Form>
+    <FormProvider {...methods}>
```
Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-form.component.tsx:62]: Passing down form contexts and  methods  
[ api](https://react-hook-form.com/docs/formprovider)

File: packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx [hunk 5]
```diff
@@ -27,34 +26,37 @@ import {
...
-  submissionNotifier,
+  isSubmittingForm,
+  setIsSubmittingForm,
+  setErrorCreating,
+  setErrorUpdating,
 }) => {
   const { t } = useTranslation();
   const { conditions, mutate } = useConditions(patientUuid);
-  const editing = formContext === 'editing';
+  const { control, watch, getValues, formState } = useFormContext<ConditionFormData>();
```
Comment (lumuchris256) [inline packages/esm-patient-conditions-app/src/conditions/conditions-widget.component.tsx:55]: using the formcontext and methods   [api ](https://react-hook-form.com/docs/useformcontext)

**Review by hadijahkyampeire — APPROVED** [review]
This looks good, let's merge it and test it more on dev3

**Pull request merged**